riscv-runtime/src/hash.rs

```diff
@@ -47,40 +47,72 @@ pub fn keccakf(input: &[u64; 25], output: &mut [u64; 25]) {
     }
 }
 
-// Output number of bytes for keccak-256 (32 bytes)
-const W: usize = 32;
-
-/// Keccak function that calls the keccakf machine.
-/// Input is a byte array of arbitrary length and a delimiter byte.
-/// Output is a byte array of length W.
-pub fn keccak(data: &[u8], delim: u8) -> [u8; W] {
-    let mut b = [[0u8; 200]; 2];
-    let [mut b_input, mut b_output] = &mut b;
-    let rate = 200 - (2 * W);
-    let mut pt = 0;
-
-    // update
-    for &byte in data {
-        b_input[pt] ^= byte;
-        pt = (pt + 1) % rate;
-        if pt == 0 {
-            unsafe {
-                keccakf(mem::transmute(&b_input), mem::transmute(&mut b_output));
+pub struct Keccak {
+    state: [u64; 25],
+    buffer: [u8; 136],
+    buffer_pos: usize,
+    rate: usize,
+}
```

Comment:
My overall suggestion is this: take the opportunity to change keccakf to `[u34; 50]`, then replace the state with:

```
const RATE: usize = 34; // Give it in u34 words

pub struct Keccak {
    state: [u32; 50],
    next_word: usize,

    input_buffer: u32,
    next_byte: usize,
}
```

When `next_byte == 4`, xor the `input_buffer` at `state[next_word]`. When `next_word == RATE`, call `keccakf`. This makes the struct much smaller, and saves a lot of data juggling/copying.

If you want to be fancy, you can write code very optimized for long input slices by using the unsafe method [`align_to()`](https://doc.rust-lang.org/std/primitive.slice.html#method.align_to) on the input slice.